src/node_v8.cc

```diff
@@ -22,23 +22,23 @@ using v8::Value;
 
 
 void GetHeapStatistics(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
+  CHECK(args.Length() == 1 && args[0]->IsObject());
+
   Isolate* isolate = args.GetIsolate();
   HeapStatistics s;
   isolate->GetHeapStatistics(&s);
-  Local<Object> info = Object::New(isolate);
-  // TODO(trevnorris): Setting many object properties in C++ is a significant
-  // performance hit. Redo this to pass the results to JS and create/set the
-  // properties there.
-#define V(name)                                                               \
-  info->Set(env->name ## _string(), Uint32::NewFromUnsigned(isolate, s.name()))
-  V(total_heap_size);
-  V(total_heap_size_executable);
-  V(total_physical_size);
-  V(used_heap_size);
-  V(heap_size_limit);
+  Local<Object> obj = args[0].As<Object>();
+  uint32_t* data =
+    static_cast<uint32_t*>(obj->GetIndexedPropertiesExternalArrayData());
```

Comment:
This should probably have a `CHECK_NE(data, nullptr)` and GetIndexedPropertiesExternalArrayDataType() and GetIndexedPropertiesExternalArrayDataLength() checks.  The last two should maybe be ASSERT_EQ checks because they are cheap but not wholly free.